File: backend/tour-design-api/src/main/java/com/tourdesign/platform/service/RecommendationHistoryService.java
package com.tourdesign.platform.service;

import com.tourdesign.platform.model.RecommendationHistoryModel;
import java.util.List;
import java.util.Optional;

public interface RecommendationHistoryService {

    List<RecommendationHistoryModel> list();

    RecommendationHistoryModel create(RecommendationHistoryModel obj);

    Optional<RecommendationHistoryModel> search(Long id);

    Optional<RecommendationHistoryModel> update(Long id, RecommendationHistoryModel obj);

    boolean delete(Long id);

    List<RecommendationHistoryModel> searchByClient(Long clientId);
}

File: frontend/tour-design-web/src/app/services/recommendation-history.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from './api.service';
import { RecommendationHistoryModel } from '../models/recommendation-history.model';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class RecommendationHistoryService {
  private baseUrl = '/recommendation-history';

  constructor(private api: ApiService) {}

  list(): Observable<RecommendationHistoryModel[]> {
    return this.api.get<RecommendationHistoryModel[]>(this.baseUrl);
  }

  search(id: number): Observable<RecommendationHistoryModel> {
    return this.api.get<RecommendationHistoryModel>(`${this.baseUrl}/${id}`);
  }

  create(obj: RecommendationHistoryModel): Observable<RecommendationHistoryModel> {
    return this.api.post<RecommendationHistoryModel>(this.baseUrl, obj);
  }

  update(id: number, obj: RecommendationHistoryModel): Observable<RecommendationHistoryModel> {
    return this.api.put<RecommendationHistoryModel>(`${this.baseUrl}/${id}`, obj);
  }

  delete(id: number): Observable<void> {
    return this.api.delete<void>(`${this.baseUrl}/${id}`);
  }

  searchByClient(clientId: number): Observable<RecommendationHistoryModel[]> {
    return this.api.get<RecommendationHistoryModel[]>(`${this.baseUrl}/searchByClient/${clientId}`);
  }
}
